src/main/host/process.rs

```diff
@@ -20,188 +20,259 @@ impl Process {
         }
     }
+}
 
-    /// Write `src` into `dst` in the current Process.
-    pub fn write_ptr_from_val<T: Copy>(&mut self, dst: PluginPtr, src: &T) -> nix::Result<()> {
-        let src = std::slice::from_ref(src);
-        self.write_ptr_from_slice(dst, src)
-    }
+pub struct ProcessMemoryReader<'a, T> {
+    process: &'a Process,
+    ptr: PluginPtr,
+    len: usize,
+    _phantom: std::marker::PhantomData<T>,
+}
 
-    /// Get a read-only slice at `src`. If memory can't be accessed, may return
-    /// an error, or may cause a Shadow error later when flushing.
+impl<'a, T> ProcessMemoryReader<'a, T>
+where
+    T: Pod,
+{
+    /// Get a read-only slice.
+    ///
+    /// This *may* have to copy the data. If you need a copy anyway use the
+    /// `std::io::Read` trait implementation avoid an extra copy.
+    ///
+    /// Fails if the source pointer is unaligned.
     ///
-    /// Prefer the `read_ptr_*` methods unless you specifically need a slice, and
-    /// the data is of non-trivial size.
-    pub fn get_slice<T>(&self, src: PluginPtr, len: usize) -> nix::Result<&[T]> {
-        if len == 0 {
+    /// Result's lifetime is intentionally bound to that of the underlying
+    /// `Process` reference rather than the lifetime of this object. That allows
+    /// this object to be discarded. e.g.:
+    ///
+    /// ```
+    /// let tv : &timeval = process.reader(ptr, 1).as_ref().unwrap()[0];
+    /// ```
+    pub fn as_ref(&self) -> Result<&'a [T], nix::errno::Errno> {
+        if self.len == 0 {
             return Ok(&[][..]);
         }
-
         // SAFETY: This never returns an invalid pointer. Passing a `p` outside
         // of the plugin's addressable memory will return a NULL pointer.
         let ptr = unsafe {
             cshadow::process_getReadablePtr(
-                self.cprocess,
-                src.into(),
-                (size_of::<T>() * len) as u64,
+                self.process.cprocess,
+                self.ptr.into(),
+                (self.len * size_of::<T>()) as u64,
             )
         } as *const T;
 
         if ptr.is_null() {
-            return Err(nix::Error::from_errno(nix::errno::Errno::EFAULT));
+            let e = nix::errno::Errno::EFAULT;
+            warn!("Reading {} bytes from {:?}: {}", self.len, self.ptr, e);
+            return Err(e);
         }
-
         if ptr.align_offset(align_of::<T>()) != 0 {
-            debug!("Unaligned pointer {:?}", ptr);
-            return Err(nix::Error::from_errno(nix::errno::Errno::EFAULT));
+            warn!("Unaligned pointer {:?}", ptr);
+            return Err(nix::errno::Errno::EFAULT);
+        }
+
+        // SAFETY: Any values are ok because T is Pod. We've validated the
+        // pointer is accessible and aligned.
+        Ok(unsafe { std::slice::from_raw_parts(ptr, self.len) })
+    }
+}
+
+impl<'a> std::io::Read for ProcessMemoryReader<'a, u8> {
+    fn read(&mut self, buf: &mut [u8]) -> std::io::Result<usize> {
```

Comment:
The individual calls *are* relatively expensive if we don't go through the `MemoryManager` path for whatever reason, especially now that we've disabled buffering by default. This is generally the expectation with `Read` objects though, so generally things that take a `Read` object should try to minimize individual `read` calls.

e.g. from https://doc.rust-lang.org/std/io/trait.Read.html:

> Please note that each call to read() may involve a system call, and therefore, using something that implements BufRead, such as BufReader, will be more efficient.

(Though I'd argue they should replace "will" with "may be", since addering a buffering layer is *less* efficient if you are accessing in large chunks)

If/when a consumer needs to do small sequential accesses, then can wrap a `Read` object in a [BufReader](https://doc.rust-lang.org/std/io/struct.BufReader.html)